core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala

```diff
@@ -202,55 +207,86 @@ private[spark] class MesosSchedulerBackend(
-        val meetsConstraints = matchesAttributeRequirements(slaveOfferConstraints, offerAttributes)
-        val meetsMemoryRequirements = mem >= calculateTotalMemory(sc)
-        val meetsCPURequirements = cpus >= (mesosExecutorCores + scheduler.CPUS_PER_TASK)
-
-        val meetsRequirements =
-          (meetsConstraints && meetsMemoryRequirements && meetsCPURequirements) ||
+  private[spark] def usableWorkerOffers(d: SchedulerDriver,
+    offers: JList[Offer]): (Seq[Protos.Offer], Seq[WorkerOffer]) = {
+    // Fail-fast on offers we know will be rejected
+    val (usableOffers, unUsableOffers) = offers.asScala.partition { o =>
+      val mem = getResource(o.getResourcesList, "mem")
+      val cpus = getResource(o.getResourcesList, "cpus")
+      val slaveId = o.getSlaveId.getValue
+      val offerAttributes = toAttributeMap(o.getAttributesList)
+
+      // check if all constraints are satisfield
+      //  1. Attribute constraints
+      //  2. Memory requirements
+      //  3. CPU requirements - need at least 1 for executor, 1 for task
+      val meetsConstraints = matchesAttributeRequirements(slaveOfferConstraints, offerAttributes)
+      val meetsMemoryRequirements = mem >= calculateTotalMemory(sc)
+      val meetsCPURequirements = cpus >= (mesosExecutorCores + scheduler.CPUS_PER_TASK)
+
+      val meetsRequirements =
+        (meetsConstraints && meetsMemoryRequirements && meetsCPURequirements) ||
           (slaveIdToExecutorInfo.contains(slaveId) && cpus >= scheduler.CPUS_PER_TASK)
 
-        // add some debug messaging
-        val debugstr = if (meetsRequirements) "Accepting" else "Declining"
-        val id = o.getId.getValue
-        logDebug(s"$debugstr offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
+      // add some debug messaging
+      val debugstr = if (meetsRequirements) "Accepting" else "Declining"
+      val id = o.getId.getValue
+      logDebug(s"$debugstr offer: $id with attributes: $offerAttributes mem: $mem cpu: $cpus")
+
+      meetsRequirements
+    }
+
+    // Decline offers we ruled out immediately
+    unUsableOffers.foreach(o => d.declineOffer(o.getId))
+
+    var availableCores = Math.max(0, maxCores - totalCoresAcquired)
 
-        meetsRequirements
+    val workerOffers = (for (o <- usableOffers) yield {
+      val coresInOffer = getResource(o.getResourcesList, "cpus")
+      val cores = if (slaveIdToExecutorInfo.contains(o.getSlaveId.getValue)) {
+        coresInOffer.toInt
+      } else {
+        // If the Mesos executor has not been started on this slave yet, set aside a few
+        // cores for the Mesos executor by offering fewer cores to the Spark executor
+        availableCores -= mesosExecutorCores
+        (coresInOffer - mesosExecutorCores).toInt
       }
 
-      // Decline offers we ruled out immediately
-      unUsableOffers.foreach(o => d.declineOffer(o.getId))
+      // check that we can still acquire cpus
+      val actualCores = Math.min(availableCores, cores).toInt
```

Comment:
Will availableCores becomes negative and just starts adding actual Cores?
I think it's probably safer to check if availableCores < Cores then we just return None.